Emit correct event user when deleting user

kimlisa opened the pull request:
#### Description
In order to capture the correct event user when deleting users, it is required to pass around contexts:
- created DeleteUser for grpc api
- updated all DeleteUser uses signature to accept context
- update unit tests

#### Screenshot
Before changes:
![Screenshot from 2020-05-13 09-23-57](https://user-images.githubusercontent.com/43280172/81854454-ced16200-9512-11ea-9785-c12ec3c2b363.png)


After changes, captures the correct user:
![screenshot](https://user-images.githubusercontent.com/43280172/81854415-be20ec00-9512-11ea-993e-86b9afcb8b30.png)

Commit:
Add DeleteUser to auth proto

Commit:
Replace REST DeleteUser with grpc

Commit:
Update unit tests

fspmarshall reviewed (commented):
GRPC method implementation looks good, but you need to take cross-version compatibility into account.  We need at least a one-version grace-period where clients can talk to old servers and servers can serve old clients.

Instead of deleting the HTTP logic, keep it around with a `DELETE IN: 5.1` (or similar) directive.  The server can simply expose HTTP and GRPC methods simultaneously.  The client should attempt a GRPC call first.  If it gets back an `UNIMPLEMENTED` error, it should fallback to the old HTTP logic.

You can find an example of compatibility fallback logic [here](https://github.com/gravitational/teleport/blob/e59afe591a371a81f3663681a12f08cec9b5cf83/lib/auth/clt.go#L1474-L1497).

alex-kovoy reviewed (commented):
Why we are replacing existing HTTP method as part of this ticket? Lets add audit events first. Plus, both methods (http, grpc) should exist side by side for 2next releases, so lets focus on events first.

Commit:
Add backwards compatibility

Commit:
Clean up comments

awly commented on lib/auth/apiserver.go:
```diff
@@ -896,12 +896,13 @@ func (s *APIServer) getUsers(auth ClientI, w http.ResponseWriter, r *http.Reques
 func (s *APIServer) deleteUser(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
 	user := p.ByName("user")
-	if err := auth.DeleteUser(user); err != nil {
+	if err := auth.DeleteUser(context.TODO(), user); err != nil {
```
use `r.Context()` here instead

kimlisa replied:
revised

awly commented on lib/auth/clt.go:
```diff
@@ -1567,10 +1567,31 @@ func (c *Client) grpcGetUsers(withSecrets bool) ([]services.User, error) {
+	}
+
+	req := &proto.DeleteUserRequest{Name: user}
+	_, grpcErr := clt.DeleteUser(ctx, req)
```
change `grpcErr` to `err` unless it needs to be different

kimlisa replied:
revised

awly reviewed (commented):
Please add a unit test for backwards-compatibility logic

Commit:
Address code reviews

Commit:
Define IClientServices interface and struct that implements it

- re-arranged code so that ClientService methods are closer together
- fixed my bug in DeleteUser that wasn't reading grpc errors correctly

Commit:
Define mocks and test backwards compatibility logic

awly commented on lib/auth/clt.go:
```diff
@@ -70,6 +70,21 @@ const (
+// Methods that are not wrappers are dependent on the roundtrip.Client instance
+// held by ClientService struct (which implements this interface).
+//
+// Interface was made for testing backward compatability logic between GRPC and REST.
```
`s/compatability/compatibility`

kimlisa replied:
revised

awly commented on lib/auth/clt.go:
```diff
@@ -70,6 +70,21 @@ const (
+// held by ClientService struct (which implements this interface).
+//
+// Interface was made for testing backward compatability logic between GRPC and REST.
+type IClientServices interface {
```
Don't prefix type names with the kind of type (like `I` here). Also, this type doesn't need to be exported.
How about `type rpcClient`? Then `ClientServices` below can be renamed to `restRPCClient`.

kimlisa replied:
revised, but had a question, is rest a type of rpc? I searched the internet and they seem to distinguish between the two

awly replied:
technically, REST is more of an API design concept.
but in practice it almost always means: "RPC over HTTP with URLs and HTTP verbs identifying the kind of operation to be performed".
so I tend to think of REST as a kind of RPC transport layer.
an alternative name here would be `httpRPCClient`

awly commented on lib/auth/clt.go:
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
+}
+
+// PostJSON is a generic method that issues http POST request to the server
+func (c *ClientServices) PostJSON(
```
don't wrap this line https://github.com/golang/go/wiki/CodeReviewComments#line-length
similarly for other methods below

kimlisa replied:
revised

awly commented on lib/auth/clt.go:
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
+	return httplib.ConvertResponse(c.Client.Delete(context.TODO(), u))
+}
+
+// Endpoint returns a URL constructed from parts and version appended, e.g.
+func (c *ClientServices) Endpoint(params ...string) string {
+	return c.Client.Endpoint(params...)
+}
```
you can skip this, since `c.Client` is embedded

awly commented on lib/auth/clt.go:
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
+//
+// The data is POSTed to HTTP server as a simple binary body (no encodings of any
+// kind are needed)
+func (c *ClientServices) PostSessionSlice(slice events.SessionSlice) error {
```
this method seems out-of-place here.
all others are generic while this is type-specific.

how about a new method `PostBytes` and another method `func (c *Client) PostSessionSlice` that calls into this?

awly commented on lib/auth/clt.go:
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	r.Header.Set("Content-Type", "application/grpc")
```
Hmm, why is this `application/grpc`?
If it's actually gRPC, should it go over the gRPC transport?

alex-kovoy replied:
@klizhentas do you recall why you had to explicitly set these headers?

awly commented on lib/auth/mock.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+// MockClient stubs the IClientServices interface.
+type MockClient struct {
```
will this get used outside of `lib/auth`?
if not, it should probably live in a `_test.go` file and be unexported.

awly commented on lib/auth/mock.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// MockAuthServiceClient stubs grpc AuthServiceClient interface.
+type MockAuthServiceClient struct {
```
same here, it should live in a `_test.go` file and be unexported.

kimlisa replied:
revised (i placed in top of `tls_test.go`)

awly commented on lib/auth/mock.go:
```diff
@@ -0,0 +1,54 @@
+
+// DeleteUser is a mock that returns an error.
+func (m *MockAuthServiceClient) DeleteUser(ctx context.Context, in *proto.DeleteUserRequest, opts ...grpc.CallOption) (*empty.Empty, error) {
+	return nil, trail.ToGRPC(trace.NotImplemented(""))
```
to make this more flexible, I suggest making the returned error a field of the struct.
that way you can experiment with other kinds of errors.

kimlisa replied:
revised

awly commented on lib/auth/tls_test.go:
```diff
@@ -2560,3 +2560,16 @@ func (s *TLSSuite) TestEventsClusterConfig(c *check.C) {
 	suite.ExpectResource(c, w, 3*time.Second, clusterNameResource)
 }
+
+func (s *TLSSuite) TestAPIBackwardsCompatibilityLogic(c *check.C) {
```
add a case where gRPC server returns a different kind of error (not `NotImplemented`)

kimlisa replied:
revised

Commit:
Revert "Define mocks and test backwards compatibility logic"

This reverts commit 11f5134c031f0f579dd7fde8515be159848d2876.

Commit:
Revert "Define IClientServices interface and struct that implements it"

This reverts commit d54836206e515ab4f69f853ffd86ce8d9d4b0950.

Commit:
Revert/Remove unnecessary interface/mock

awly commented on lib/auth/tls_test.go:
```diff
@@ -45,13 +47,35 @@ import (
+// MockAuthServiceClient stubs grpc AuthServiceClient interface.
+type mockAuthServiceClient struct {
+	proto.AuthServiceClient
+	errType error
```
nit: how about `err` or `deleteUserErr`?

kimlisa replied:
revised to `err`

Commit:
Address CR and rebase
